File: templates/ruflet_flutter_template/apple_packages/ruflet_apple/Sources/RufletExtensions/RufletMap/Sources/PolygonLayer.swift
import RufletEngine
import SwiftUI

struct RufletPolygonDescriptor {
  let control: RufletControl
  let coordinates: [RufletMapCoordinate]
  let borderWidth: Double
  let borderColor: Color
  let fillColor: Color
  let label: String?
  let labelStyle: RufletTextStyle?
  let rotatesLabel: Bool
  let lineCap: CGLineCap
  let lineJoin: CGLineJoin
  let polygonCulling: Bool
  let polygonLabels: Bool
  let drawLabelsLast: Bool
  let simplificationTolerance: Double
  let useAlternativeRendering: Bool
}

@MainActor
func rufletPolygonDescriptors(_ layer: RufletControl) -> [RufletPolygonDescriptor] {
  let polygonCulling = layer.boolean("polygon_culling", default: true)
  let polygonLabels = layer.boolean("polygon_labels", default: true)
  let drawLabelsLast = layer.boolean("draw_labels_last", default: false)
  let simplificationTolerance = layer.number("simplification_tolerance", default: 0.3) ?? 0.3
  let useAlternativeRendering = layer.boolean("use_alternative_rendering", default: false)
  return layer.children("polygons", visibleOnly: false).compactMap {
    polygon -> RufletPolygonDescriptor? in
    guard polygon.type == "PolygonMarker" else { return nil }
    let points = rufletMapCoordinates(polygon.value("coordinates"))
    guard points.count >= 3 else { return nil }
    polygon.notifyParent = true
    return RufletPolygonDescriptor(
      control: polygon,
      coordinates: points,
      borderWidth: polygon.number("border_stroke_width", default: 0) ?? 0,
      borderColor: rufletMapColor(polygon.string("border_color"), default: .green),
      fillColor: rufletMapColor(polygon.string("color"), default: .green),
      label: polygon.string("label"),
      labelStyle: parseTextStyle(polygon.value("label_text_style")),
      rotatesLabel: polygon.boolean("rotate_label", default: false),
      lineCap: rufletLineCap(polygon.string("stroke_cap")),
      lineJoin: rufletLineJoin(polygon.string("stroke_join")),
      polygonCulling: polygonCulling,
      polygonLabels: polygonLabels,
      drawLabelsLast: drawLabelsLast,
      simplificationTolerance: simplificationTolerance,
      useAlternativeRendering: useAlternativeRendering)
  }
}

struct PolygonLayerControl: View {
  @ObservedObject var control: RufletControl
  var body: some View { EmptyView() }
}

File: templates/ruflet_flutter_template/ruflet_packages/ruflet_map/lib/src/polygon_layer.dart
import 'package:ruflet/ruflet.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_map/flutter_map.dart';

import 'utils/map.dart';

class PolygonLayerControl extends StatelessWidget with RufletStoreMixin {
  final Control control;

  const PolygonLayerControl({super.key, required this.control});

  @override
  Widget build(BuildContext context) {
    debugPrint("PolygonLayerControl build: ${control.id}");

    var polygons = control
        .children("polygons")
        .where((c) => c.type == "PolygonMarker")
        .map((polygon) {
      polygon.notifyParent = true;
      return Polygon(
          borderStrokeWidth: polygon.getDouble("border_stroke_width", 0)!,
          borderColor: polygon.getColor(
              "border_color", context, const Color(0xFF4CAF50))!,
          color: polygon.getColor("color", context, const Color(0xFF4CAF50))!,
          disableHolesBorder: polygon.getBool("disable_holes_border", false)!,
          rotateLabel: polygon.getBool("rotate_label", false)!,
          label: polygon.getString("label"),
          labelStyle: polygon.getTextStyle("label_text_style",
              RufletStyleTheme.of(context), const TextStyle())!,
          strokeCap: polygon.getStrokeCap("stroke_cap", StrokeCap.round)!,
          strokeJoin: polygon.getStrokeJoin("stroke_join", StrokeJoin.round)!,
          points: polygon
              .get("coordinates", [])!
              .map((c) => parseLatLng(c))
              .nonNulls
              .toList());
    }).toList();

    return PolygonLayer(
      polygons: polygons,
      polygonCulling: control.getBool("polygon_culling", true)!,
      polygonLabels: control.getBool("polygon_labels", true)!,
      drawLabelsLast: control.getBool("draw_labels_last", false)!,
      simplificationTolerance:
          control.getDouble("simplification_tolerance", 0.3)!,
      useAltRendering: control.getBool("use_alternative_rendering", false)!,
    );
  }
}
